python/cudf/cudf/core/frame.py

```diff
@@ -458,6 +458,61 @@ def reassign_categories(categories, cols, col_idxs):
 
         return out
 
+    def equals(self, other, **kwargs):
+        """
+        Test whether two objects contain the same elements.
+        This function allows two Series or DataFrames to be compared against
+        each other to see if they have the same shape and elements. NaNs in
+        the same location are considered equal. The column headers do not
+        need to have the same type.
+
+        Parameters
+        ----------
+        other : Series or DataFrame
+            The other Series or DataFrame to be compared with the first.
+
+        Returns
+        -------
+        bool
+            True if all elements are the same in both objects, False
+            otherwise.
+
```

Comment:
```suggestion
        Returns
        -------
        bool
            True if all elements are the same in both objects, False
            otherwise.

        Notes
        -----
        This API returns `True` when the underlying data are
        of different `dtypes` but have the same values.
```

Leaving a note to differentiate our behaviour from pandas.

Reply:
Do we want to differentiate from Pandas in this case?

Reply:
Yes

```python
>>> import cudf
>>> df = cudf.DataFrame({1: [10], 2: [20]})
>>> different_data_type = cudf.DataFrame({1: [10.0], 2: [20.0]})
>>> df.equals(different_data_type)
True
>>> df.to_pandas().equals(different_data_type.to_pandas())
False
```

Reply:
FWIW, in `cudf` we don't check the column labels for equality at all. So `cudf.DataFrame({'a': []}).equals(cudf.DataFrame({'b': []}))`.

Since column labels are just Pandas objects, we can always check for equality of columns with `self.columns.equals(other.columns)` in `DataFrame.equals()`. Should we?

FWIW, Series names aren't compared for equality...

Reply:
We could also loop over the column names in the ColumnAccessor and compare them with `==`.

Reply:
> Since column labels are just Pandas objects, we can always check for equality of columns with `self.columns.equals(other.columns)` in `DataFrame.equals()`. Should we?

What is Pandas behavior? If they check the column names then we should as well.

Reply:
👍 They do

Reply:
What about dtype equality? Is that something we want to enforce? @kkraus14 @galipremsagar?

Reply:
Looks like numpy does not have dtype equality check but pyarrow does. 

```python
>>> import numpy as np
>>> x = np.array([1, 2, 3])
>>> y = x.astype('float64')
>>> x
array([1, 2, 3])
>>> y
array([1., 2., 3.])
>>> np.array_equal(x, y)
True

>>> import pyarrow as pa
>>> x1 = pa.array(x)
>>> y1 = pa.array(y)
>>> x1
<pyarrow.lib.Int64Array object at 0x11d245468>
[
  1,
  2,
  3
]
>>> y1
<pyarrow.lib.DoubleArray object at 0x10fa84a08>
[
  1,
  2,
  3
]
>>> x1.equals(y1)
False
>>> y1.equals(x1)
False
>>> y1 = pa.array([1, 2, 3])
>>> x1.equals(y1)
True
```

I think it would be ideal to have dtype equality checked too since both pandas and pyarrow are performing that.

Reply:
Thanks @galipremsagar. Added both `dtype` and `name` equality checks and tests. Note that names are checked only for DataFrames and not for Series/Index.